aws-lc-rs/src/rsa/key.rs

```diff
@@ -146,126 +136,54 @@ impl KeyPair {
     /// * `Unspecified`: Any key generation failure.
     #[cfg(feature = "fips")]
     pub fn generate_fips(size: KeySize) -> Result<Self, Unspecified> {
-        let private_key = generate_rsa_key(size.bit_len(), true)?;
-        Self::new(private_key).map_err(|_| Unspecified)
+        let private_key = generate_rsa_key(size.bits(), true)?;
+        Ok(Self::new(private_key)?)
     }
 
-    /// Parses an unencrypted PKCS#8-encoded RSA private key.
-    ///
-    /// A RSA keypair may be generated using [`KeyPair::generate`].
-    ///
-    /// Only two-prime (not multi-prime) keys are supported. The public modulus
-    /// (n) must be at least 2047 bits. The public modulus must be no larger
-    /// than 4096 bits. It is recommended that the public modulus be exactly
-    /// 2048 or 3072 bits. The public exponent must be at least 65537.
-    ///
-    /// The following will generate a 2048-bit RSA private key of the correct form using
-    /// OpenSSL's command line tool:
-    ///
-    /// ```sh
-    ///    openssl genpkey -algorithm RSA \
-    ///        -pkeyopt rsa_keygen_bits:2048 \
-    ///        -pkeyopt rsa_keygen_pubexp:65537 | \
-    ///      openssl pkcs8 -topk8 -nocrypt -outform der > rsa-2048-private-key.pk8
-    /// ```
-    ///
-    /// The following will generate a 3072-bit RSA private key of the correct form:
+    /// Parses an unencrypted PKCS#8 DER encoded RSA private key.
     ///
-    /// ```sh
-    ///    openssl genpkey -algorithm RSA \
-    ///        -pkeyopt rsa_keygen_bits:3072 \
-    ///        -pkeyopt rsa_keygen_pubexp:65537 | \
-    ///      openssl pkcs8 -topk8 -nocrypt -outform der > rsa-3072-private-key.pk8
-    /// ```
+    /// Keys can be generated using [`KeyPair::generate`].
     ///
-    /// Often, keys generated for use in OpenSSL-based software are stored in
-    /// the Base64 “PEM” format without the PKCS#8 wrapper. Such keys can be
-    /// converted to binary PKCS#8 form using the OpenSSL command line tool like
-    /// this:
+    /// # *ring*-compatibility
     ///
-    /// ```sh
-    /// openssl pkcs8 -topk8 -nocrypt -outform der \
-    ///     -in rsa-2048-private-key.pem > rsa-2048-private-key.pk8
-    /// ```
-    ///
-    /// Base64 (“PEM”) PKCS#8-encoded keys can be converted to the binary PKCS#8
-    /// form like this:
-    ///
-    /// ```sh
-    /// openssl pkcs8 -nocrypt -outform der \
-    ///     -in rsa-2048-private-key.pem > rsa-2048-private-key.pk8
-    /// ```
+    /// *aws-lc-rs* does not impose the same limitations that *ring* does for
+    /// RSA keys. Thus signatures may be generated by keys that are not accepted
+    /// by *ring*. In particular:
+    /// * RSA private keys ranging between 2048-bit keys and 8192-bit keys are supported.
+    /// * The public exponent does not have a required minimum size.
```

Comment:
Can this generate end up calling the FIPS version or is that a different Rust public API? If so the FIPS check does set a minimum/max for the public exponent.

Reply:
There is a separate `generate_fips` which calls `RSA_generate_key_fips` in AWS-LC, and the Rust `generate` calls `RSA_generate_key_ex` in AWS-LC instead, which explicitly sets doesn't has the FIPS flow path disabled (`check_fips=0`) in the AWS-LC implementation.